flax/ids.py

```diff
@@ -0,0 +1,43 @@
+# Copyright 2022 The Flax Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""UUIDs for Flax internals."""
+
+import threading
+
+"""
+We need globally unique key ids for Module and Variable object instances
+to preserve and recreate sharing-by-reference relationship when lifting
+transforms and adopting outside Modules.
+ - Use of id() is unacceptable because these identifiers are literally pointers
+   which can be recycled, so we rely on a globally unique counter id instead.
+ - We avoid the use of the random builtin uuid.uuid4() mainly to preserve
+   the deterministic execution of flax internals.
+"""
+
+
+class UUIDManager:
+  """Globally unique counter-based id manager."""
+  def __init__(self):
+    self._lock = threading.Lock()
+    self._id = 0
+
+  def __call__(self):
+    with self._lock:
+      self._id += 1
+      cur_id = self._id
+    return cur_id
```

Comment:
Does this work as well?

```suggestion
    with self._lock:
      self._id += 1
    return self._id
```

Reply:
I'm probably being overly paranoid due to old C programming trauma and weird lock/mutex edge conditions around what is supposed to be in the critical section, but I just tried to break this in python and I couldn't, so I think the simpler form is ok - I replaced it.

Reply:
thinking about it again and looking at some python bytecode disassembly -- technically the simpler version could be wrong if a thread switch occurs between exiting the with-block and re-accessing the value of _id before returning.... so even though it's an edge cases and we don't even really want to support multithreaded module construction, I'd rather do the correct thing.

Reply:
Makes sense, returning directly seems like the best option.